crates/bevy_reflect/src/reflect.rs

```diff
@@ -0,0 +1,77 @@
+use crate::{serde::Serializable, List, Map, Struct, TupleStruct};
+use std::{any::Any, fmt::Debug};
+
+pub use bevy_utils::AHasher as ReflectHasher;
+
+pub enum ReflectRef<'a> {
+    Struct(&'a dyn Struct),
+    TupleStruct(&'a dyn TupleStruct),
+    List(&'a dyn List),
+    Map(&'a dyn Map),
+    Value(&'a dyn Reflect),
+}
+
+pub enum ReflectMut<'a> {
+    Struct(&'a mut dyn Struct),
+    TupleStruct(&'a mut dyn TupleStruct),
+    List(&'a mut dyn List),
+    Map(&'a mut dyn Map),
+    Value(&'a mut dyn Reflect),
+}
+
+/// A reflected rust type.
+pub trait Reflect: Any + Send + Sync {
+    fn type_name(&self) -> &str;
+    fn any(&self) -> &dyn Any;
+    fn any_mut(&mut self) -> &mut dyn Any;
+    fn apply(&mut self, value: &dyn Reflect);
+    fn set(&mut self, value: Box<dyn Reflect>) -> Result<(), Box<dyn Reflect>>;
+    fn reflect_ref(&self) -> ReflectRef;
+    fn reflect_mut(&mut self) -> ReflectMut;
+    fn clone_value(&self) -> Box<dyn Reflect>;
+    /// Returns a hash of the value (which includes the type) if hashing is supported. Otherwise `None` will be returned.
+    fn hash(&self) -> Option<u64>;
+    /// Returns a "partial equal" comparison result if comparison is supported. Otherwise `None` will be returned.
+    fn partial_eq(&self, _value: &dyn Reflect) -> Option<bool>;
+    /// Returns a serializable value, if serialization is supported. Otherwise `None` will be returned.
+    fn serializable(&self) -> Option<Serializable>;
+}
+
+impl Debug for dyn Reflect {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&format!("Reflect({})", self.type_name()))
+    }
+}
+
+impl dyn Reflect {
+    pub fn downcast<T: Reflect>(self: Box<dyn Reflect>) -> Result<Box<T>, Box<dyn Reflect>> {
+        // SAFE?: Same approach used by std::any::Box::downcast. ReflectValue is always Any and type has been checked.
+        if self.is::<T>() {
```

Comment:
Yea, this should be fine, and you'll be in good company ([example](https://github.com/hyperium/hyper/blob/a470446deb2cb2c0e3700f67d9f70097d0d7d75f/src/upgrade.rs#L244-L253))